### DISCUSSION TITLE
How to run faiss add_index on a selected GPU in a multi GPU system ?

### DISCUSSION THREAD

**Discussion by shamanez** [facebookresearch/faiss#3638] [General]
My training process uses 4 32-GB GPUs and during the training process I want to update the index. There I have add_index function to run. But I want to run only on the 5th GPU. Is there a way that I can select the exact GPU?

Running on:
- [ ] CPU
- [X ] GPU

Interface: 
- [ ] C++
- [ X] Python

**Comment by wickedfoo** [comment]
Specify the `device` in the GpuIndexConfig passed to the index constructor. Different GPU indices have different config objects (GpuIndexIVFPQConfig, GpuIndexFlatConfig, etc) but they all have `device` as a member.

e.g.,

https://github.com/facebookresearch/faiss/blob/efa1e3f64fe6256d6ede1334a60af4d91f1c9e71/benchs/kmeans_mnist.py#L63

devices are numbered from 0 (following the CUDA convention).

**Comment by shamanez** [comment]
@wickedfoo 

Thanks a lot.  Can you please explain to me what is happening in the this [script](https://github.com/facebookresearch/faiss/wiki/Running-on-GPUs#in-python-1)?

```
# build a flat (CPU) index
index_flat = faiss.IndexFlatL2(d)
# make it into a gpu index
gpu_index_flat = faiss.index_cpu_to_gpu(res, 0, index_flat)
```

What is the second parameter? is it GPU index? if not how can use a specific GPU in this case?
